### PR TITLE
[SPARK-38473][CORE][COMMON][DOCS][SQL] Use error classes in org.apache.spark.scheduler

### PR DESCRIPTION
**Pull request by asmitalim**
Issue: https://issues.apache.org/jira/browse/SPARK-38473

### What changes were proposed in this pull request?

1. Addition of new error classes to org.apache.spark.scheduler
2. Changing the scheduler tests to catch the new exceptions produced by these classes
3. Formatting changes, as a result of running ./dev/scalafmt (should I revert these?)
4. Changes to InMemoryBaseTable.scala, to prevent breaking of sbt doc (https://github.com/apache/spark/pull/43874)

### Why are the changes needed?

To move exceptions in the package org.apache.scheduler to use error classes

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Running existing tests and updating existing tests to catch the new exceptions and check the error class: IE, SparkIllegalStateException is caught instead of IllegalStateException.

### Was this patch authored or co-authored using generative AI tooling?

No.

Please note: the sqlState fields in the error classes need to be confirmed with someone, I would appreciate your input on this!

### CODE REVIEW

**Comment by hannahkamundson** [comment]
@MaxGekk @bozhang2820  Hello :) I know you are both busy so no pressure but would you happen to have time to review this PR?
https://issues.apache.org/jira/browse/SPARK-38473

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -527,6 +527,12 @@
     ],
     "sqlState" : "22018"
   },
+  "COURSE_GRAINED_SCHEDULER_NEGATIVE_EXECUTORS" : {
+    "message" : [
+      "Attempted to request a negative number of additional executor(s) <numAdditionalExecutors> from the cluster manager. Please specify a positive number!"
+    ],
+    "sqlState" : "42K0J"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:534]: @srielau WDYT of the sql state?

**Comment by mridulm** [comment]
Most of the diff appears to be formatting change, can you clean that up please ? Thanks

**Comment by asmitalim** [comment]
> Most of the diff appears to be formatting change, can you clean that up please ? Thanks

On it! Those were generated as a result of running ./dev/scalafmt before pushing changes. I'm reverting them.

File: core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala [hunk 2]
```diff
@@ -136,7 +139,9 @@ private class AsyncEventQueue(
    */
   private[scheduler] def stop(): Unit = {
     if (!started.get()) {
-      throw new IllegalStateException(s"Attempted to stop $name that has not yet started!")
+      throw new SparkIllegalStateException(
+        errorClass = "SCHEDULER_ASYNC_STOP",
+        messageParameters = Map("name" -> name.toString))
     }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala:145]: We can avoid these sort of off error classes - particularly for cases which are not due to user input, but due to internal state of spark.

Comment (asmitalim) [reply core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala:145]: Do you suggest using an INTERNAL_ERROR_SCHEDULER class for this instead? There are similar INTERNAL_ERROR classes used for core

Comment (asmitalim) [reply core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala:145]: Hi @mridulm should I do this?

**Pull request closed without merging**